sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala

```diff
@@ -246,85 +260,238 @@ object SpecifiedWindowFrame {
+    extends OffsetWindowFunction {
 
-  override def dataType: DataType = throw new UnresolvedException(this, "dataType")
-  override def foldable: Boolean = throw new UnresolvedException(this, "foldable")
-  override def nullable: Boolean = throw new UnresolvedException(this, "nullable")
-  override lazy val resolved = false
+  def this(input: Expression, offset: Expression) = this(input, offset, Literal(null))
 
-  override def init(): Unit = throw new UnresolvedException(this, "init")
-  override def reset(): Unit = throw new UnresolvedException(this, "reset")
-  override def prepareInputParameters(input: InternalRow): AnyRef =
-    throw new UnresolvedException(this, "prepareInputParameters")
-  override def update(input: AnyRef): Unit = throw new UnresolvedException(this, "update")
-  override def batchUpdate(inputs: Array[AnyRef]): Unit =
-    throw new UnresolvedException(this, "batchUpdate")
-  override def evaluate(): Unit = throw new UnresolvedException(this, "evaluate")
-  override def get(index: Int): Any = throw new UnresolvedException(this, "get")
+  def this(input: Expression) = this(input, Literal(1))
 
-  override def toString: String = s"'$name(${children.mkString(",")})"
+  def this() = this(Literal(null))
 
-  override def newInstance(): WindowFunction = throw new UnresolvedException(this, "newInstance")
+  val offsetSign = 1
 }
 
-case class UnresolvedWindowExpression(
-    child: UnresolvedWindowFunction,
-    windowSpec: WindowSpecReference) extends UnaryExpression with Unevaluable {
+case class Lag(input: Expression, offset: Expression, default: Expression)
+    extends OffsetWindowFunction {
 
-  override def dataType: DataType = throw new UnresolvedException(this, "dataType")
-  override def foldable: Boolean = throw new UnresolvedException(this, "foldable")
-  override def nullable: Boolean = throw new UnresolvedException(this, "nullable")
-  override lazy val resolved = false
-}
+  def this(input: Expression, offset: Expression) = this(input, offset, Literal(null))
 
-case class WindowExpression(
-    windowFunction: WindowFunction,
-    windowSpec: WindowSpecDefinition) extends Expression with Unevaluable {
+  def this(input: Expression) = this(input, Literal(1))
 
-  override def children: Seq[Expression] = windowFunction :: windowSpec :: Nil
+  def this() = this(Literal(null))
 
-  override def dataType: DataType = windowFunction.dataType
-  override def foldable: Boolean = windowFunction.foldable
-  override def nullable: Boolean = windowFunction.nullable
+  val offsetSign = -1
+}
 
-  override def toString: String = s"$windowFunction $windowSpec"
+abstract class AggregateWindowFunction extends DeclarativeAggregate with WindowFunction {
+  self: Product =>
+  override val frame = SpecifiedWindowFrame(RowFrame, UnboundedPreceding, CurrentRow)
+  override def dataType: DataType = IntegerType
+  override def nullable: Boolean = false
+  override val mergeExpressions = Nil // TODO how to deal with this?
```

Comment:
There is a `supportsPartial` field in `AggregateFunction`. We can override that and set it as `false`? Also, if `mergeExpressions` is not expected to be called, we can throw an `UnsupportedOperationException`.